### PR TITLE
Refactor genetic_algorithm example

### PR DESCRIPTION
**Pull request by megmcca**
This PR refactors the genetic_algorithm library example to handle some common errors, add some functionality, and prepare the genetic algorithm for parallelization. The core code of libmod_optimize.py has not been changed.

The changes are broadly as follows:

- the genetic_algorithm_SNAP and genetic_algorithm_ACE folders have been combined into one (tested), and the two READMEs from those folders have been consolidated
- the running of the initial fit and initiating the chosen optimization algorithm has been moved to a new script, script_optimize.py. This script uses the classes and functions from libmod_optimize.py as a module
- a latin hypercube sampling scheme (written by @jmgoff) has been added and is what populates the initial generation of the genetic algorithm
- a new function “prep_fitsnap_input" has been added to libmod_optimize.py. It attempts to handle several cases which may crash the current version of this example or lead to unexpected behavior (for example, fitting to stresses, smartweights, and a common group folder error "(low<=high)" )
- an optional parameter in the genetic_algorithm class, "write_to_json," has been added, which prints out a standard JSON file with all the input settings used in a genetic algorithm fit with the term "fs_input.json" appended to the user's chosen potential file name setting. It is turned off by default, but is toggled on in script_optimize.py.
- a directory "1Aug23_Standard" with two very short fits (popsize, ngenerations = 4, 4) has been added to show the expected output 

Warnings (to be updated):
- The current version requires at least 4 groups to be defined (will let user know). 
- The current version of the genetic algorithm does not work for fitting to stresses
- The current version of the genetic algorithm automatically turns off smartweights, but the user can turn them back on with the "smartweights_override" argument to genetic_algorithm (this is not recommended!)
- It is not yet known how the current version handles a train/test split and any random_sampling included in it. It is currently recommended to set training_size to 1 and testing_size to 0 (or simply, 'size' to 1).

Many thanks to the primary author James Goff @jmgoff for help and for the initial code.

### CODE REVIEW

File: examples/library/genetic_algorithm/1Aug23_Standard/Ta-ACE_metrics.md [hunk 1]
```diff
@@ -1,54 +1,54 @@
...
-| ('Volume_BCC', 'Unweighted', 'Training', 'Energy')    |       21 | 0.0167189   |   0.0218718   |      0.999991    |
-| ('Volume_BCC', 'Unweighted', 'Training', 'Force')     |      126 | 3.67219e-14 |   7.2479e-14  | -13175.4         |
-| ('Volume_BCC', 'weighted', 'Training', 'Energy')      |       21 | 0.00911094  |   0.011919    |      0.999991    |
-| ('Volume_BCC', 'weighted', 'Training', 'Force')       |        0 | 0           | nan           |    nan           |
-| ('Volume_FCC', 'Unweighted', 'Training', 'Energy')    |       31 | 0.00609592  |   0.00755758  |      0.999997    |
-| ('Volume_FCC', 'Unweighted', 'Training', 'Force')     |      372 | 4.03487e-14 |   9.78894e-14 |     -2.84276e+06 |
-| ('Volume_FCC', 'weighted', 'Training', 'Energy')      |       31 | 0.005192    |   0.00643692  |      0.999997    |
-| ('Volume_FCC', 'weighted', 'Training', 'Force')       |        0 | 0           | nan           |    nan           |
\ No newline at end of file
+|                                                       |   ncount |             mae |            rmse |              rsq |
```
Comment (rohskopf) [inline examples/library/genetic_algorithm/1Aug23_Standard/Ta-ACE_metrics.md:1]: The errors in this file are much worse than previously, was a different objective function used or something?

If so we can make a new standard.

Comment (megmcca) [reply examples/library/genetic_algorithm/1Aug23_Standard/Ta-ACE_metrics.md:1]: yes, this is expected and was due to my settings - the fits are based on a test-sized population and generation setting (only 4 and 4), so pretty terrible!

it was just to confirm that things were running

i'll re-run both ACE and SNAP with the population and generation values used in the previous examples and update the Standards directory

File: examples/library/genetic_algorithm/libmod_optimize.py
[the remark this answers is not in the document]
Comment (rohskopf) [reply examples/library/genetic_algorithm/libmod_optimize.py:1]: Removing SA is good unless others use it or if there's some other advantage. Definitely keep it around somewhere.

I hope we can abstract enough to just pass a set of parameters into whatever optimization algorithm. Best to get this situated with the GA first.

[the remark this answers is not in the document]
Comment (megmcca) [reply examples/library/genetic_algorithm/libmod_optimize.py:1]: good suggestions, thanks James!

i've got most of them in now (plus a couple other little bug fixes):

1. parameters that were inside of the module are now exposed to the user as arguments for the GA in script_optimize.py
1 a) user-defined force_delta_keywords list implemented and working
1 b) my_w_ranges is now an input parameter. i agree about implementing the same for forces and stresses soon, but had a couple questions for you (we'll talk soon)
1 c) etot_weight and ftot_weight have also been made into input params

2. the new `conv_thr = 1e-10` was borrowed from another example script you gave me, so not mine actually ;-) and yes, seems to perform well so let's keep it

3. the new flag with `check_gen` is now in there, but in testing, i'm seeing significantly longer convergence times with multiple variations on `population_size` and `ngenerations.` for some reason. i've commented it out for now and will do a little more testing 

4. about removing the SA - in general, I agree with Drew, the SA and the GA share many similar classes/functions, so already pretty nicely abstracted.
i guess one way to decide would be to ask - do those shared classes/functions need to be refined/debugged themselves? if so, then it makes sense to keep the SA and GA together for now at least

anyway, tldr; most is done, just parts of 1 b) and 3 left to do; will update soon

[the remark this answers is not in the document]
Comment (jmgoff) [reply examples/library/genetic_algorithm/libmod_optimize.py:1]: 1b) sounds good!
3) A little scary that it is taking a lot longer (I'm hoping it is for good reasons only haha) - a couple things we can check here:
-  A best score/vs generation plot to see if it is running after convergence (fluctuating about the minimum with values larger than conv_thr) If that is the case, we just need to make conv_thr less stringent. Implementing some sort of adaptive threshold may help here too. 
- Interplay between initial population (latin hypercube vs random), initial population size, and the # of iterations to convergence. I would suspect random initial populations and/or smaller initial populations with LH sampling are more likely to converge to a local minimum rather than a global one. If we converge faster, but to a less good /local minimum, we still want to avoid faster convergence in that case.
- If it seems like it is something else, I/we can try to define a convergence criteria class with a bunch of different functions so we can test against other criteria systematically.

4) good points from you and @rohskopf  about the SA - maybe other algorithms like EGO could be used later if we talk to some subject matter experts 😄

**Comment by megmcca** [comment]
after a bit of discussion, James and I think this is ready to go 
we'll put all new changes into a fresh PR

**Review by rohskopf — APPROVED** [review]
Clean update to the GA script, thanks @megmcca and @jmgoff
